src/dune_pkg/rev_store.ml

```diff
@@ -2,43 +2,97 @@ open Stdune
 open Dune_vcs
 module Process = Dune_engine.Process
 module Display = Dune_engine.Display
+module Scheduler = Dune_engine.Scheduler
 module Re = Dune_re
+module Flock = Dune_util.Flock
 open Fiber.O
 
-type t = { dir : Path.t }
+type t =
+  { dir : Path.t
+  ; lock : Path.t
+  }
+
 type rev = Rev of string
 
-let equal { dir } t = Path.equal dir t.dir
+(* Async-inspired variation of [Fiber.repeat_until] *)
+let rec repeat_until_finished state f =
+  let* computation = f state in
+  match computation with
+  | `Repeat state -> repeat_until_finished state f
+  | `Finished result -> Fiber.return result
+;;
+
+let lock_limited flock lock timeout =
+  let timeout = max timeout 0. in
+  let sleep_duration = 0.1 in
+  repeat_until_finished timeout (fun timeout ->
+    if timeout > 0.
+    then (
+      match Flock.lock_non_block flock lock with
+      | Ok `Success as ok -> Fiber.return @@ `Finished ok
+      | Ok `Failure ->
+        let+ () = Scheduler.sleep sleep_duration in
+        `Repeat (timeout -. sleep_duration)
+      | err -> Fiber.return @@ `Finished err)
+    else Fiber.return @@ `Finished (Ok `Failure))
+;;
+
+let with_flock lock_path ~f =
+  let open Fiber.O in
+  let parent = Path.parent_exn lock_path in
+  Path.mkdir_p parent;
+  let fd = Unix.openfile (Path.to_string lock_path) [ Unix.O_CREAT; O_RDONLY ] 0o644 in
+  let flock = Flock.create fd in
+  let max_duration = 5. in
+  Fiber.finalize
+    ~finally:(fun () ->
+      (* closing the fd releases the flock automatically *)
+      Unix.close fd;
+      (* delete the lock to signal to the user we don't hold a lock *)
+      Fiber.return @@ Path.unlink_no_err lock_path)
+    (fun () ->
+      let* acquired = lock_limited flock Flock.Exclusive max_duration in
+      match acquired with
+      | Ok `Success -> f ()
+      | Ok `Failure ->
+        Code_error.raise
+          (sprintf "Couldn't acquire lock after %.1f seconds" max_duration)
+          []
+      | Error error ->
+        Code_error.raise "lock" [ "error", Dyn.string (Unix.error_message error) ])
```

Comment:
I think we just want to reraise it as `Unix.Unix_error error`. This is what will happen on filesystems that don't support flock. I suppose NFS is going to throw errors here.
Not strictly needed but you can test that by injecting an error with `strace`, I did that in a sendfile test (grepping for `sendfile` in the test suite should find it).

Reply:
A unix error is better than a code error, but a user error with the context would be even better.

Reply:
Ah yes, the `sendfile` trick is pretty cool, I wasn't aware that `strace` can also inject failures. But I guess it makes sense if it intercepts all syscalls :+1: